src/analysis/inner_product_space/reproducing_kernel.lean

```diff
@@ -0,0 +1,256 @@
+/-
+Copyright (c) 2022 Shing Tak Lam. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Shing Tak Lam
+-/
+
+import analysis.inner_product_space.projection
+import analysis.inner_product_space.positive
+
+/-!
+# Reproducing Kernel Hilbert Space
+
+A Hilbert space `H` is called a reproducing kernel Hilbert space (RKHS) if it is a vector subspace
+of `X → V` where `V` is a Hilbert space, and the evaluation map `H → V` at each `X` is continuous.
+
+## Main definitions
+
+* `rkhs 𝕜 X V H` - `H` is a RKHS, where `𝕜` is the scalar field, `X` is the domain, `V` is the
+  Hilbert space.
+* `rkhs.kernel` - the kernel of a RKHS, as a continuous linear map `V →L[𝕜] V`
+* `rkhs.eval'` - For `V = 𝕜`, by the Riesz representation theorem, the evaluation map can be
+  represented as the inner product with a point in `H`, which we call the `eval'`
+* `rkhs.scalar_kernel` - For `V = 𝕜`, we can represent the kernel of `H` as an inner product with a
+  point in `H`, which we call the `scalar_kernel`
+* `rkhs.eval'_span` - The span of the `eval'`s as a subspace of `H`.
+
+## Main statements
+
+* `rkhs.eval'_span_topological_closure_eq_top` - the topological closude of the span of the
+  `eval's` is the whole space `H`.
+* `rkhs.tendsto_nhds_apply` - Convergence in `H` implies pointwise convergence.
+
+## Notes
+
+Due to the statement of the Riesz representation theorem in mathlib, the order of `x` and `y` in
+the definition of the `scalar_kernel` and the representation of it as an inner product of
+`eval'`s is reversed. I opted to keep the statement simpler, but this leaves the order
+reversed.
+
+## Tags
+
+Reproducing kernel Hilbert space, RKHS
+-/
+
+noncomputable theory
+
+open_locale inner_product topological_space
+open inner_product_space filter
+
+variables (𝕜 : Type*) [is_R_or_C 𝕜]
+variables (X : Type*)
+variables (V : Type*) [inner_product_space 𝕜 V]
+
+/--
+A reproducing kernel Hilbert space (RKHS) is a vector subspace of `X → V` where evaluation at a
+point is continuous.
+-/
+class rkhs (H : Type*) [inner_product_space 𝕜 H] extends fun_like H X (λ _, V) :=
```

Comment:
The linter is not happy with this, since in the `fun_like` instance `𝕜` is not used. Not sure what to do here.

Reply:
We were just discussing [here](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/normed_field/near/297808567) about the possibility of unbundling the analysis classes from the algebraic classes. If that were the case, could you just define an `rkhs` as something that takes `H : submodule 𝕜 (X → V)`, as an argument, and then extends `inner_product_space 𝕜 H` and has a field `continuous_eval`?

Reply:
I can define it to be a predicate on a `submodule 𝕜 (X → V)` as well. The issue with that approach is that

```lean
example (H : submodule 𝕜 (X → V)) : topological_space H := infer_instance
```

works, since it has the subspace topology of the prod topology. If I define an inner product space instance on `H` then there would be two distinct topologies. I'm not sure how unbundling the topological structure and the algebraic structure will help here, since it's just picking up the `subtype` topology.

So I suppose what I need is a way to have a _vector_ subspace of `X → V` that isn't a _topological_ subspace of `X → V`?

Reply:
What about using a type synonym, like `algebraic_submodule` (terrible name) where you only copy over the algebraic structure? This would incur a bit of overhead, but I think it wouldn't be so bad once you get past the basic API.

Reply:
@shingtaklam1324 I haven't checked this first idea, so it may not work, but what about just temporarily disabling the `topological_space` instance on functions. For example, can you just do `local attribute [-instance] Pi.topological_space`. Then you might be able to use an `submodule 𝕜 (X → V)`.

As a second option which I'm fairly certain will work: instead of having an *actual* submodule or extending `fun_like`, just use an injective linear map, like this
```lean
class rkhs (H : Type*) [inner_product_space 𝕜 H] :=
(to_funs : H →ₗ[𝕜] (X → V))
(to_funs_injective : function.injective (to_funs : H → X → V))
(continuous_eval' : ∀ (x : X), continuous (λ (f : H), to_funs f x))
```
You can then set up a `fun_like` instance in the natural way. I think if the first one doesn't work, then the second is definitely the way to go unless I'm missing something.